File: packages/shared-types/src/domain/absence.ts
// Absence domain types - Core business entity

import type {
  TenantEntity,
  UUID,
  ISODateString,
  ISODateTimeString,
} from "../utils/common";
import type { DateRange, DayPortion, Duration } from "../utils/dates";

/** Absence type categories */
export type AbsenceCategory =
  | "planned" // Vacations, RTT, planned leave
  | "unplanned" // Sick leave, emergency
  | "statutory"; // Maternity, paternity, etc.

/** Standard absence types */
export type AbsenceType =
  | "paid_leave" // Congés payés
  | "rtt" // RTT (French specific)
  | "sick_leave" // Arrêt maladie
  | "sick_leave_workplace" // Accident du travail
  | "maternity" // Congé maternité
  | "paternity" // Congé paternité
  | "parental" // Congé parental
  | "bereavement" // Congé décès
  | "wedding" // Congé mariage
  | "moving" // Congé déménagement
  | "unpaid_leave" // Congé sans solde
  | "training" // Formation
  | "remote_work" // Télétravail (tracking only)
  | "other"; // Autre

/** Absence status workflow */
export type AbsenceStatus =
  | "draft" // Created but not submitted
  | "pending" // Awaiting approval
  | "approved" // Approved by manager
  | "rejected" // Rejected
  | "cancelled" // Cancelled by employee
  | "completed"; // Past absence

/** Absence entity */
export interface Absence extends TenantEntity {
  /** Employee taking the absence */
  employeeId: UUID;
  /** Absence type */
  type: AbsenceType;
  /** Category */
  category: AbsenceCategory;
  /** Start date */
  startDate: ISODateString;
  /** End date */
  endDate: ISODateString;
  /** Start day portion */
  startPortion: DayPortion;
  /** End day portion */
  endPortion: DayPortion;
  /** Duration calculation */
  duration: Duration;
  /** Status */
  status: AbsenceStatus;
  /** Employee comment/reason */
  reason?: string;
  /** Manager comment */
  managerComment?: string;
  /** Rejection reason */
  rejectionReason?: string;
  /** Approver (manager) */
  approverId?: UUID;
  /** Approved/Rejected timestamp */
  decisionAt?: ISODateTimeString;
  /** Medical certificate required */
  medicalCertificateRequired: boolean;
  /** Medical certificate uploaded */
  medicalCertificateUploaded: boolean;
  /** Replacement employee (if assigned) */
  replacementEmployeeId?: UUID;
  /** Source system (for imports) */
  sourceSystem?: string;
  /** External ID (for sync) */
  externalId?: string;
  /** Recurrence pattern (for recurring absences) */
  recurrencePattern?: RecurrencePattern;
}

/** Recurrence pattern for recurring absences */
export interface RecurrencePattern {
  frequency: "daily" | "weekly" | "monthly";
  interval: number;
  daysOfWeek?: number[]; // ISO: 1=Monday
  endDate?: ISODateString;
  occurrences?: number;
}

/** Absence request (for creation/update) */
export interface AbsenceRequest {
  employeeId: UUID;
  type: AbsenceType;
  startDate: ISODateString;
  endDate: ISODateString;
  startPortion?: DayPortion;
  endPortion?: DayPortion;
  reason?: string;
  recurrencePattern?: RecurrencePattern;
}

/** Absence summary for listings */
export interface AbsenceSummary {
  id: UUID;
  employeeId: UUID;
  employeeName: string;
  type: AbsenceType;
  category: AbsenceCategory;
  startDate: ISODateString;
  endDate: ISODateString;
  duration: Duration;
  status: AbsenceStatus;
}

/** Absence with employee details (for manager views) */
export interface AbsenceWithEmployee extends Absence {
  employee: {
    id: UUID;
    displayName: string;
    email: string;
    jobTitle: string;
    departmentId: UUID;
    departmentName: string;
  };
}

/** Daily absence summary (for calendar views) */
export interface DailyAbsenceSummary {
  date: ISODateString;
  totalAbsent: number;
  byType: Record<AbsenceType, number>;
  byDepartment: Record<UUID, number>;
  absences: AbsenceSummary[];
}

/** Absence statistics */
export interface AbsenceStatistics {
  period: DateRange;
  totalDays: number;
  totalWorkingDays: number;
  byType: Record<AbsenceType, AbsenceTypeStats>;
  byCategory: Record<AbsenceCategory, number>;
  byDepartment: Record<UUID, DepartmentAbsenceStats>;
  trends: AbsenceTrend[];
}

/** Stats per absence type */
export interface AbsenceTypeStats {
  count: number;
  totalDays: number;
  avgDuration: number;
  percentOfTotal: number;
}

/** Department absence statistics */
export interface DepartmentAbsenceStats {
  departmentId: UUID;
  departmentName: string;
  headcount: number;
  totalAbsenceDays: number;
  absenceRate: number; // percentage
  byType: Record<AbsenceType, number>;
}

/** Absence trend data point */
export interface AbsenceTrend {
  period: string; // YYYY-MM or YYYY-WXX
  absenceRate: number;
  totalDays: number;
  byCategory: Record<AbsenceCategory, number>;
}

/** Absence conflict (overlapping absences) */
export interface AbsenceConflict {
  absenceId: UUID;
  conflictingAbsenceId: UUID;
  overlapDays: number;
  overlapRange: DateRange;
}

/** Absence calendar event (for UI) */
export interface AbsenceCalendarEvent {
  id: UUID;
  title: string;
  start: ISODateString;
  end: ISODateString;
  type: AbsenceType;
  status: AbsenceStatus;
  employeeId: UUID;
  employeeName: string;
  color?: string;
}

File: apps/api/app/models/absence.py
"""Absence model — core business entity for leave tracking.

Maps to shared-types: Absence, AbsenceType, AbsenceCategory, AbsenceStatus,
DayPortion.
"""

import enum
import uuid
from datetime import date, datetime

from sqlalchemy import Date, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, TenantMixin, sa_enum


class AbsenceType(str, enum.Enum):
    PAID_LEAVE = "paid_leave"
    RTT = "rtt"
    SICK_LEAVE = "sick_leave"
    SICK_LEAVE_WORKPLACE = "sick_leave_workplace"
    MATERNITY = "maternity"
    PATERNITY = "paternity"
    PARENTAL = "parental"
    BEREAVEMENT = "bereavement"
    WEDDING = "wedding"
    MOVING = "moving"
    UNPAID_LEAVE = "unpaid_leave"
    TRAINING = "training"
    REMOTE_WORK = "remote_work"
    OTHER = "other"


class AbsenceCategory(str, enum.Enum):
    PLANNED = "planned"
    UNPLANNED = "unplanned"
    STATUTORY = "statutory"


class AbsenceStatus(str, enum.Enum):
    DRAFT = "draft"
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"
    CANCELLED = "cancelled"
    COMPLETED = "completed"


class DayPortion(str, enum.Enum):
    FULL = "full"
    MORNING = "morning"
    AFTERNOON = "afternoon"


class Absence(TenantMixin, Base):
    """Absence record.

    Tracks leave periods for employees with full audit trail
    (who approved, when, with what reason).
    """

    __tablename__ = "absences"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )
    organization_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("organizations.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    employee_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("employees.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    type: Mapped[AbsenceType] = mapped_column(sa_enum(AbsenceType), nullable=False)
    category: Mapped[AbsenceCategory] = mapped_column(
        sa_enum(AbsenceCategory), nullable=False
    )
    start_date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    end_date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    start_portion: Mapped[DayPortion] = mapped_column(
        sa_enum(DayPortion), default=DayPortion.FULL
    )
    end_portion: Mapped[DayPortion] = mapped_column(
        sa_enum(DayPortion), default=DayPortion.FULL
    )
    # Duration as JSONB (days, workingDays, calendarDays)
    duration: Mapped[dict] = mapped_column(  # type: ignore[type-arg]
        JSONB, nullable=False, server_default="{}"
    )
    business_days: Mapped[int] = mapped_column(Integer, default=0)
    status: Mapped[AbsenceStatus] = mapped_column(
        sa_enum(AbsenceStatus), default=AbsenceStatus.DRAFT
    )
    reason: Mapped[str | None] = mapped_column(Text)
    manager_comment: Mapped[str | None] = mapped_column(Text)
    rejection_reason: Mapped[str | None] = mapped_column(Text)
    approver_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
    )
    decision_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    medical_certificate_required: Mapped[bool] = mapped_column(default=False)
    medical_certificate_uploaded: Mapped[bool] = mapped_column(default=False)
    replacement_employee_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("employees.id", ondelete="SET NULL"),
    )
    source_system: Mapped[str | None] = mapped_column(String(100))
    external_id: Mapped[str | None] = mapped_column(String(255))
    # Recurrence pattern (RecurrencePattern from shared-types)
    recurrence_pattern: Mapped[dict | None] = mapped_column(JSONB)  # type: ignore[type-arg]

    def __repr__(self) -> str:
        return f"<Absence {self.type.value} {self.start_date}–{self.end_date}>"
